pkg/helios/cas/cas.go

```diff
@@ -0,0 +1,398 @@
+	
+	// Process all hashes first (CPU-bound, can be optimized)
+	for i, content := range contents {
+		// Get hasher from pool for zero allocation
+		hasher := s.hasherPool.Get().(*blake3.Hasher)
+		hasher.Write(content)
+		digest := hasher.Sum(nil)
+		hasher.Reset()
+		s.hasherPool.Put(hasher)
+		
+		hashes[i] = types.Hash{
+			Algorithm: types.BLAKE3,
+			Digest:    digest,
+		}
+		hashKeys[i] = fmt.Sprintf("%x", digest)
+	}
+	
+	// Batch cache operations (single lock acquisition)
+	s.mutex.Lock()
+	for i, content := range contents {
+		hashKey := hashKeys[i]
+		if _, exists := s.cache[hashKey]; !exists {
+			s.cache[hashKey] = make([]byte, len(content))
+			copy(s.cache[hashKey], content)
+		}
+	}
+	s.mutex.Unlock()
+	
+	// Batch key cache operations
+	s.keyMutex.Lock()
+	for i, hash := range hashes {
+		digestStr := string(hash.Digest)
+		s.keyCache[digestStr] = hashKeys[i]
+	}
+	s.keyMutex.Unlock()
+	
+	// Handle persistence based on mode
+	if !s.memoryMode {
+		// Queue all writes together for better batching
+		for i, content := range contents {
+			filePath := filepath.Join(s.storePath, hashKeys[i])
+			
+			// Critical: Add to WaitGroup BEFORE attempting to send to channel
+			s.wg.Add(1)
+			
+			select {
+			case s.writeQueue <- writeOp{filePath: filePath, content: content}:
+				// Successfully queued
+			default:
+				// Queue full, must call Done() since we already called Add()
+				s.wg.Done()
+				// Fallback to sync write
+				os.WriteFile(filePath, content, 0644) // Ignore error in batch mode
+			}
+		}
+	}
+	
+	return hashes, nil
+}
+
+// Load retrieves content by its hash
+func (s *BLAKE3Store) Load(hash types.Hash) ([]byte, error) {
+	// Validate hash algorithm
+	if hash.Algorithm != types.BLAKE3 {
+		return nil, fmt.Errorf("unsupported hash algorithm: %s", hash.Algorithm)
+	}
+
+	// Fast path: check key cache first
+	digestStr := string(hash.Digest)
+	s.keyMutex.RLock()
+	hashKey, keyExists := s.keyCache[digestStr]
+	s.keyMutex.RUnlock()
+	
+	if !keyExists {
+		hashKey = fmt.Sprintf("%x", hash.Digest)
+		// Cache the key for future lookups
+		s.keyMutex.Lock()
+		s.keyCache[digestStr] = hashKey
+		s.keyMutex.Unlock()
+	}
+
+	// Try cache first (L1 cache hit)
+	s.mutex.RLock()
+	if content, exists := s.cache[hashKey]; exists {
+		// Return copy to prevent external modification
+		result := make([]byte, len(content))
+		copy(result, content)
+		s.mutex.RUnlock()
+		return result, nil
+	}
+	s.mutex.RUnlock()
+
+	// Load from disk (L2 storage)
+	filePath := filepath.Join(s.storePath, hashKey)
+	content, err := os.ReadFile(filePath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, fmt.Errorf("content not found for hash %s", hashKey)
+		}
+		return nil, fmt.Errorf("failed to read content from disk: %w", err)
+	}
+
+	// Cache for future access
+	s.mutex.Lock()
+	s.cache[hashKey] = make([]byte, len(content))
+	copy(s.cache[hashKey], content)
+	s.mutex.Unlock()
+
+	return content, nil
```

Comment:
Return-value consistency: cache hits return a copy, but disk path returns the original slice. Consider returning a copy in both cases (or document the difference) to avoid surprising callers that mutate the returned slice.